plugins/out_s3/s3.c

```diff
@@ -1735,21 +1808,19 @@ static void s3_upload_queue(struct flb_config *config, void *out_context)
             ctx->upload_queue_success = FLB_FALSE;
 
             /* If retry limit was reached, discard file and remove file from queue */
-            upload_contents->retry_counter++;
-            if (upload_contents->retry_counter >= MAX_UPLOAD_ERRORS) {
-                flb_plg_warn(ctx->ins, "Chunk file failed to send %d times, will not "
-                             "retry", upload_contents->retry_counter);
+            if (upload_contents->upload_file->failures > ctx->ins->retry_limit) {
                 s3_store_file_inactive(ctx, upload_contents->upload_file);
                 multipart_upload_destroy(upload_contents->m_upload_file);
                 remove_from_queue(upload_contents);
                 continue;
```

Comment:
One thing here... I think the `upload_data` function will already remove the s3_file... please double check this won't cause a double free issue when the upload queue code also removes the s3_file

Reply:
Did you check this?